File: cobalt/tools/performance/memory/verify_memory_trace_unittest.cc
// Copyright 2026 The Cobalt Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "build/build_config.h"

#if (BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_MAC)) && !BUILDFLAG(IS_STARBOARD)

#include "base/command_line.h"
#include "base/files/file_path.h"
#include "base/files/file_util.h"
#include "base/files/scoped_temp_dir.h"
#include "base/path_service.h"
#include "base/process/launch.h"
#include "testing/gtest/include/gtest/gtest.h"

namespace cobalt {
namespace {

base::FilePath GetMemoryToolsDir() {
  base::FilePath source_root;
  EXPECT_TRUE(
      base::PathService::Get(base::DIR_SRC_TEST_DATA_ROOT, &source_root));
  return source_root.AppendASCII("cobalt")
      .AppendASCII("tools")
      .AppendASCII("performance")
      .AppendASCII("memory");
}

TEST(VerifyMemoryTraceTest, CleanTraceProceeds) {
  base::ScopedTempDir temp_dir;
  ASSERT_TRUE(temp_dir.CreateUniqueTempDir());

  base::FilePath trace_path =
      temp_dir.GetPath().AppendASCII("clean_trace.json");
  std::string trace_content = R"({
    "traceEvents": [
      {
        "cat": "__metadata",
        "name": "process_name",
        "args": {"name": "Cobalt"}
      },
      {
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
          "dumps": {
            "heaps_v2": {"allocators": {"malloc": {}}},
            "process_mmaps": {"vm_regions": []},
            "process_totals": {
              "private_footprint_bytes": "5025000",
              "peak_resident_set_size": "a5b6000"
            }
          }
        }
      }
    ]
  })";
  ASSERT_TRUE(base::WriteFile(trace_path, trace_content));

  base::FilePath script_path =
      GetMemoryToolsDir().AppendASCII("verify_memory_trace.py");

  base::CommandLine cmd(base::FilePath(FILE_PATH_LITERAL("python3")));
  cmd.AppendArgPath(script_path);
  cmd.AppendArgPath(trace_path);

  std::string output;
  int exit_code = -1;
  bool success = base::GetAppOutputWithExitCode(cmd, &output, &exit_code);

  EXPECT_TRUE(success) << "Failed to run verify_memory_trace.py.";
  EXPECT_EQ(0, exit_code) << "Script failed with output:\n" << output;
  EXPECT_NE(output.find("[SIGNAL: PROCEED]"), std::string::npos)
      << "Expected [SIGNAL: PROCEED] in output:\n"
      << output;
  EXPECT_NE(output.find("Private: 80.1 MB"), std::string::npos)
      << "Expected parsed Private Footprint in scorecard:\n"
      << output;
}

TEST(VerifyMemoryTraceTest, NoisyTraceWarns) {
  base::ScopedTempDir temp_dir;
  ASSERT_TRUE(temp_dir.CreateUniqueTempDir());

  base::FilePath trace_path =
      temp_dir.GetPath().AppendASCII("noisy_trace.json");
  std::string trace_content = R"({
    "traceEvents": [
      {
        "cat": "toplevel,blink",
        "name": "TaskQueue::RunTask"
      },
      {
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
          "dumps": {
            "heaps_v2": {"allocators": {}},
            "process_mmaps": {"vm_regions": []}
          }
        }
      }
    ]
  })";
  ASSERT_TRUE(base::WriteFile(trace_path, trace_content));

  base::FilePath script_path =
      GetMemoryToolsDir().AppendASCII("verify_memory_trace.py");

  base::CommandLine cmd(base::FilePath(FILE_PATH_LITERAL("python3")));
  cmd.AppendArgPath(script_path);
  cmd.AppendArgPath(trace_path);

  std::string output;
  int exit_code = -1;
  bool success = base::GetAppOutputWithExitCode(cmd, &output, &exit_code);

  EXPECT_TRUE(success) << "Failed to run verify_memory_trace.py.";
  EXPECT_EQ(0, exit_code) << "Script failed with output:\n" << output;
  EXPECT_NE(output.find("[SIGNAL: PROCEED_WITH_WARNING]"), std::string::npos)
      << "Expected [SIGNAL: PROCEED_WITH_WARNING] in output:\n"
      << output;
  EXPECT_NE(output.find("🔴 NOISY (blink, toplevel)"), std::string::npos)
      << "Expected noise categories in scorecard:\n"
      << output;
}

TEST(VerifyMemoryTraceTest, IncompleteTraceAborts) {
  base::ScopedTempDir temp_dir;
  ASSERT_TRUE(temp_dir.CreateUniqueTempDir());

  base::FilePath trace_path =
      temp_dir.GetPath().AppendASCII("incomplete_trace.json");
  std::string trace_content = R"({
    "traceEvents": [
      {
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
          "dumps": {
            "process_mmaps": {"vm_regions": []}
          }
        }
      }
    ]
  })";
  ASSERT_TRUE(base::WriteFile(trace_path, trace_content));

  base::FilePath script_path =
      GetMemoryToolsDir().AppendASCII("verify_memory_trace.py");

  base::CommandLine cmd(base::FilePath(FILE_PATH_LITERAL("python3")));
  cmd.AppendArgPath(script_path);
  cmd.AppendArgPath(trace_path);

  std::string output;
  int exit_code = -1;
  bool success = base::GetAppOutputWithExitCode(cmd, &output, &exit_code);

  EXPECT_TRUE(success) << "Failed to run verify_memory_trace.py.";
  EXPECT_EQ(1, exit_code) << "Expected exit code 1 for incomplete trace.";
  EXPECT_NE(output.find("[SIGNAL: ABORT]"), std::string::npos)
      << "Expected [SIGNAL: ABORT] in output:\n"
      << output;
}

TEST(VerifyMemoryTraceTest, MissingProcessMmapsAborts) {
  base::ScopedTempDir temp_dir;
  ASSERT_TRUE(temp_dir.CreateUniqueTempDir());

  base::FilePath trace_path =
      temp_dir.GetPath().AppendASCII("missing_mmaps.json");
  std::string trace_content = R"({
    "traceEvents": [
      {
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
          "dumps": {
            "heaps_v2": {"allocators": {}}
          }
        }
      }
    ]
  })";
  ASSERT_TRUE(base::WriteFile(trace_path, trace_content));

  base::FilePath script_path =
      GetMemoryToolsDir().AppendASCII("verify_memory_trace.py");

  base::CommandLine cmd(base::FilePath(FILE_PATH_LITERAL("python3")));
  cmd.AppendArgPath(script_path);
  cmd.AppendArgPath(trace_path);

  std::string output;
  int exit_code = -1;
  bool success = base::GetAppOutputWithExitCode(cmd, &output, &exit_code);

  EXPECT_TRUE(success);
  EXPECT_EQ(1, exit_code);
  EXPECT_NE(output.find("[SIGNAL: ABORT]"), std::string::npos);
}

TEST(VerifyMemoryTraceTest, CorruptedJsonAborts) {
  base::ScopedTempDir temp_dir;
  ASSERT_TRUE(temp_dir.CreateUniqueTempDir());

  base::FilePath trace_path =
      temp_dir.GetPath().AppendASCII("corrupted_trace.json");
  std::string trace_content = "{ this is invalid json ::: ";
  ASSERT_TRUE(base::WriteFile(trace_path, trace_content));

  base::FilePath script_path =
      GetMemoryToolsDir().AppendASCII("verify_memory_trace.py");

  base::CommandLine cmd(base::FilePath(FILE_PATH_LITERAL("python3")));
  cmd.AppendArgPath(script_path);
  cmd.AppendArgPath(trace_path);

  std::string output;
  int exit_code = -1;
  bool success = base::GetAppOutputWithExitCode(cmd, &output, &exit_code);

  EXPECT_TRUE(success);
  EXPECT_EQ(1, exit_code);
  EXPECT_NE(output.find("[SIGNAL: ABORT]"), std::string::npos);
}

TEST(VerifyMemoryTraceTest, NonExistentFileAborts) {
  base::ScopedTempDir temp_dir;
  ASSERT_TRUE(temp_dir.CreateUniqueTempDir());

  base::FilePath trace_path =
      temp_dir.GetPath().AppendASCII("non_existent_file.json");

  base::FilePath script_path =
      GetMemoryToolsDir().AppendASCII("verify_memory_trace.py");

  base::CommandLine cmd(base::FilePath(FILE_PATH_LITERAL("python3")));
  cmd.AppendArgPath(script_path);
  cmd.AppendArgPath(trace_path);

  std::string output;
  int exit_code = -1;
  bool success = base::GetAppOutputWithExitCode(cmd, &output, &exit_code);

  EXPECT_TRUE(success);
  EXPECT_EQ(1, exit_code);
  EXPECT_NE(output.find("[SIGNAL: ABORT]"), std::string::npos);
}

}  // namespace
}  // namespace cobalt

#endif  // (BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_MAC)) &&
        // !BUILDFLAG(IS_STARBOARD)

File: cobalt/tools/performance/memory/verify_memory_trace_unittest.py
#!/usr/bin/env python3
# Copyright 2026 The Cobalt Authors. All Rights Reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
"""Unit tests for verify_memory_trace.py."""

import json
import os
import sys
import tempfile
import unittest

script_dir = os.path.dirname(os.path.realpath(__file__))
if script_dir not in sys.path:
  sys.path.insert(0, script_dir)

# pylint: disable=wrong-import-position
import verify_memory_trace


class VerifyMemoryTraceTest(unittest.TestCase):
  """Test suite for memory trace verification and noise detection."""

  def setUp(self):
    # pylint: disable=consider-using-with
    self.temp_dir = tempfile.TemporaryDirectory()

  def tearDown(self):
    self.temp_dir.cleanup()

  def test_clean_memory_trace(self):
    """Verifies that a pure memory trace triggers the PROCEED signal."""
    clean_events = [
        {
            "cat": "__metadata",
            "name": "process_name",
            "args": {
                "name": "Cobalt"
            }
        },
        {
            "cat": "disabled-by-default-memory-infra",
            "name": "periodic_interval",
            "args": {
                "dumps": {
                    "heaps_v2": {
                        "allocators": {
                            "malloc": {}
                        }
                    },
                    "process_mmaps": {
                        "vm_regions": []
                    },
                }
            },
        },
    ]
    metrics = verify_memory_trace.analyze_memory_events(clean_events)
    self.assertEqual(metrics["memory_dumps_count"], 1)
    self.assertTrue(metrics["has_heaps_v2"])
    self.assertTrue(metrics["has_process_mmaps"])
    self.assertFalse(metrics["has_redundant_data"])
    self.assertEqual(metrics["noise_categories"], set())
    self.assertTrue(verify_memory_trace.print_verdict(metrics))

  def test_clean_memory_trace_with_process_totals(self):
    """Verifies parsing of process_totals (Private Footprint and Peak RSS)."""
    events = [{
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
            "dumps": {
                "heaps_v2": {},
                "process_mmaps": {},
                "process_totals": {
                    "private_footprint_bytes": "5025000",
                    "peak_resident_set_size": "a5b6000",
                    "is_peak_rss_resettable": True,
                },
            }
        },
    }]
    metrics = verify_memory_trace.analyze_memory_events(events)
    self.assertTrue(metrics["has_process_totals"])
    self.assertAlmostEqual(
        metrics["latest_private_footprint_mb"], 80.14, places=1)
    self.assertAlmostEqual(metrics["latest_peak_rss_mb"], 165.71, places=1)

  def test_parse_bytes_to_mb(self):
    """Verifies byte parsing logic and type guards."""
    # pylint: disable=protected-access
    self.assertIsNone(verify_memory_trace._parse_bytes_to_mb(None))
    self.assertIsNone(verify_memory_trace._parse_bytes_to_mb(True))
    self.assertIsNone(verify_memory_trace._parse_bytes_to_mb(False))
    self.assertIsNone(verify_memory_trace._parse_bytes_to_mb("invalid_hex"))
    self.assertAlmostEqual(
        verify_memory_trace._parse_bytes_to_mb(1048576), 1.0, places=4)
    self.assertAlmostEqual(
        verify_memory_trace._parse_bytes_to_mb("100000"), 1.0, places=4)

  def test_noisy_memory_trace(self):
    """Verifies that CPU categories trigger PROCEED_WITH_WARNING."""
    noisy_events = [
        {
            "cat": "__metadata",
            "name": "process_name"
        },
        {
            "cat": "toplevel,blink",
            "name": "TaskQueue::RunTask"
        },
        {
            "cat": "netlog",
            "name": "URLRequest"
        },
        {
            "cat": "disabled-by-default-memory-infra",
            "name": "periodic_interval",
            "args": {
                "dumps": {
                    "heaps_v2": {
                        "allocators": {
                            "malloc": {}
                        }
                    },
                    "process_mmaps": {
                        "vm_regions": []
                    },
                }
            },
        },
    ]
    metrics = verify_memory_trace.analyze_memory_events(noisy_events)
    self.assertEqual(metrics["memory_dumps_count"], 1)
    self.assertTrue(metrics["has_heaps_v2"])
    self.assertTrue(metrics["has_process_mmaps"])
    self.assertTrue(metrics["has_redundant_data"])
    self.assertEqual(metrics["noise_categories"],
                     {"toplevel", "blink", "netlog"})
    # Should still return True (profileable), but flagged as noisy
    self.assertTrue(verify_memory_trace.print_verdict(metrics))

  def test_incomplete_trace_missing_heaps(self):
    """Verifies that missing heaps_v2 triggers ABORT."""
    incomplete_events = [{
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
            "dumps": {
                "process_mmaps": {
                    "vm_regions": []
                }
            }
        },
    }]
    metrics = verify_memory_trace.analyze_memory_events(incomplete_events)
    self.assertFalse(metrics["has_heaps_v2"])
    self.assertFalse(verify_memory_trace.print_verdict(metrics))

  def test_incomplete_trace_missing_mmaps(self):
    """Verifies that missing process_mmaps triggers ABORT."""
    incomplete_events = [{
        "cat": "disabled-by-default-memory-infra",
        "name": "periodic_interval",
        "args": {
            "dumps": {
                "heaps_v2": {
                    "allocators": {}
                }
            }
        },
    }]
    metrics = verify_memory_trace.analyze_memory_events(incomplete_events)
    self.assertFalse(metrics["has_process_mmaps"])
    self.assertFalse(verify_memory_trace.print_verdict(metrics))

  def test_parse_valid_json_file(self):
    """Tests parsing a standard trace file."""
    trace_path = os.path.join(self.temp_dir.name, "trace.json")
    trace_content = {
        "traceEvents": [{
            "cat": "__metadata",
            "name": "process_name"
        }, {
            "cat": "disabled-by-default-memory-infra",
            "name": "periodic_interval",
            "args": {
                "dumps": {
                    "heaps_v2": {},
                    "process_mmaps": {}
                }
            }
        }]
    }
    with open(trace_path, "w", encoding="utf-8") as f:
      json.dump(trace_content, f)

    events = verify_memory_trace.parse_trace_file(trace_path)
    self.assertIsNotNone(events)
    self.assertEqual(len(events), 2)
    self.assertTrue(verify_memory_trace.verify_trace(trace_path))

  def test_parse_missing_file(self):
    """Tests parsing a non-existent file."""
    non_existent = os.path.join(self.temp_dir.name, "missing.json")
    self.assertIsNone(verify_memory_trace.parse_trace_file(non_existent))
    self.assertFalse(verify_memory_trace.verify_trace(non_existent))


if __name__ == "__main__":
  unittest.main()
